flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricFetcherTest.java

```diff
@@ -184,4 +193,52 @@ public String getValue() {
 
 		return dump;
 	}
+
+	@Test
+	public void testLongUpdateInterval() {
+		final long updateInterval = 1000L;
+		final AtomicInteger requestMetricQueryServicePathsCounter = new AtomicInteger(0);
+		final RestfulGateway restfulGateway = createRestfulGateway(requestMetricQueryServicePathsCounter);
+
+		final MetricFetcher<RestfulGateway> fetcher = createMetricFetcher(updateInterval, restfulGateway);
+
+		fetcher.update();
+		fetcher.update();
+
+		assertThat(requestMetricQueryServicePathsCounter.get(), is(1));
+	}
+
+	@Test
+	public void testShortUpdateInterval() throws InterruptedException {
+		final long updateInterval = 1L;
+		final AtomicInteger requestMetricQueryServicePathsCounter = new AtomicInteger(0);
+		final RestfulGateway restfulGateway = createRestfulGateway(requestMetricQueryServicePathsCounter);
+
+		final MetricFetcher<RestfulGateway> fetcher = createMetricFetcher(updateInterval, restfulGateway);
+
+		fetcher.update();
+		Thread.sleep(2L * updateInterval);
+		fetcher.update();
+
+		assertThat(requestMetricQueryServicePathsCounter.get(), is(2));
+	}
+
+	@Nonnull
+	private MetricFetcher<RestfulGateway> createMetricFetcher(long updateInterval, RestfulGateway restfulGateway) {
+		return new MetricFetcher<>(
+			() -> CompletableFuture.completedFuture(restfulGateway),
+			path -> new CompletableFuture<>(),
```

Comment:
Returning an already completed future should be a safer default implementation.

Reply:
I actually think the opposite is true, because you would not expect that if you get a response, then complete this future that all subsequent responses are completed yet.